### ISSUE TITLE
A persistent failure of one route should not impact the ability to send through another

### ISSUE THREAD

**Issue by neilisfragile** [matrix-org/sygnal#104]
So if Sygnal cannot send iOS pushes, but Android is still operational, then Android pushes should continue to work.

**Comment by richvdh** [comment]
Afaik there are no queues currently.

**Comment by richvdh** [comment]
@neilisfragile could I suggest phrasing this issue in terms of the problem that we're trying to solve?

**Comment by Half-Shot** [comment]
I'm surprised that a failure to send one notification would lead to another being impacted, at least on a small scale. Is this related to us having too many open connections to the IOS API and therefore impacting Android's ability to work?

**Comment by richvdh** [comment]
@Half-Shot: well, empirically all push died despite only one certificate expiring.

We've raised a task to investigate in more detail about what happened.

**Comment by clokep** [comment]
Looking at this briefly, it seems that if a notification dispatch for a particular device fails then subsequent notifications will be skipped

The exceptions for `dispatch_notification` are caught outside the loop instead of per device, see:

https://github.com/matrix-org/sygnal/blob/8567224dd5112599cbe0a997aaccc075d7b5c254/sygnal/http.py#L215-L266

It seems this is done so that the response can be properly set to an error if the notification fails, this seems buggy though since a notification could partially go through and still get an error response. (E.g. if the initial device is successful and the second device is a failure...) Unfortunately this doesn't really seem to explain what is seen here though since it seems Synapse only sends a single device at a time:

https://github.com/matrix-org/synapse/blob/0d775fcc2d0c7b6a07dad5430256d4d6c75a9f0d/synapse/push/httppusher.py#L309-L370

From this it isn't clear to me how there could be an interaction between different pushkins.

**Comment by richvdh** [comment]
It's very frustrating that we haven't really looked at this until 11 days later, by which time all the logs have rotated. I think we need to reconsider how this sort of investigation gets prioritised.

Having said that, we have some information from our various prometheus metrics.

Note that we have two separate sygnal instances. These are routed to by haproxy, which has slightly different configurations for the two servers: `sygnal-twisted-1` has a connection limit of 200, while `sygnal-twisted-0` is unlimited.

First `sygnal-twisted-1`. At 14:10 UTC, we see connections stacking up on haproxy, where they time out after 5 seconds:

![image](https://user-images.githubusercontent.com/1389908/80001650-821bdd80-84b6-11ea-96c4-0e2e133bce4e.png)

This is consistent with sygnal suddenly taking a long time to process all or some requests, and is enough to explain why pushes unrelated to the failed certificate started failing.

Some related graphs from the sygnal end showing the same thing:

![image](https://user-images.githubusercontent.com/1389908/80002680-d7a4ba00-84b7-11ea-8296-82cab56317f8.png)
![image](https://user-images.githubusercontent.com/1389908/80002747-ef7c3e00-84b7-11ea-8803-484ab44d9212.png)
![image](https://user-images.githubusercontent.com/1389908/80002771-fb680000-84b7-11ea-983e-ad132ae30cc2.png)

... so basically all the APNS requests start stacking up infinitely. Meanwhile the queue time for GCM pushes drops to zero, presumably because haproxy starts dropping a percentage of them:

![image](https://user-images.githubusercontent.com/1389908/80003111-729d9400-84b8-11ea-8273-100ca2756342.png)


----

ok, so now `sygnal-twisted-0`. This seemed to be fine until the sygnal process was restarted at 18:11 UTC, when suddenly lots of requests start timing out. Presumably the cert is only checked when we reconnect to the APNS server, or something.

![image](https://user-images.githubusercontent.com/1389908/80008645-ec854b80-84bf-11ea-89d0-2a9ae8e12c1b.png)

The 5xxs are initially due to haproxy timing out the request (after 3 minutes), or giving up trying to connect to the server (after 10 retries) when we start getting connection errors later:

![image](https://user-images.githubusercontent.com/1389908/80016157-a1bd0100-84ca-11ea-9350-939a17a4c653.png)

The request timeouts appear to be the same as before: that Sygnal starts tarpitting the requests. I think the connection errors are just due to Sygnal taking a long time to accept, possibly because it was busy garbage-collecting:

![image](https://user-images.githubusercontent.com/1389908/80017136-03319f80-84cc-11ea-8a51-e792ef706d4b.png)

It looks like the percentage of non-APNS requests actually failing on this server was much smaller, since they weren't being blocked by the haproxy.

---

So in summary the behaviour can be explained by
 * when the APNS requests failed, sygnal started taking a very long time to process those requests
 * that meant that the slots on the haproxy got full of APNS requests
 * so other requests started failing too.

So the main question is why sygnal wasn't responding to these failing requests, and instead was allowing them to stack up forever, leading to 5GB of memory use.

Some other operational questions:
 * should we increase the maxconn on sygnal-twisted-1? I seem to think we added it because we had problems with large numbers of requests stacking up when things got busy
 * shouldn't sygnal-twisted-0 have a maxconn?

**Comment by richvdh** [comment]
> So the main question is why sygnal wasn't responding to these failing requests, and instead was allowing them to stack up forever, leading to 5GB of memory use.

Part of the problem here might be the retry schedule:

 * 4 attempts at 1 second intervals in [aioapns](https://github.com/Fatal1ty/aioapns/blob/0587f33/aioapns/connection.py#L366-L373) (we set `pool.max_connection_attempts=3`, but I think there is an off-by-one error. The connection pool stuff doesn't really do anything until a connection is established.)
 * 3 attempts at 10 and 20 second intervals in [apnspushkin](https://github.com/matrix-org/sygnal/blob/v0.4.1/sygnal/apnspushkin.py#L230-#L251).

So that's a total of `3s+10s+3s+20s+3s == 39s` for each push, so that doesn't really explain why it kept going up, rather than plateauing. 

I'm at a bit of a loss here. I think we'd have to set up a test sygnal and try to reproduce the problem somehow, which seems a bit out of scope here.

**Comment by richvdh** [comment]
> So that's a total of 3s+10s+3s+20s+3s == 39s for each push, so that doesn't really explain why it kept going up, rather than plateauing.

... although it is of course enough time to explain why the server with a maxconn soon got saturated with apns requests.

**Comment by richvdh** [comment]
ok I'm going to mark this as done in favour of #116 and #117
